test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp

```diff
@@ -1,54 +1,133 @@
+ * @param non_random_generator_instance Callable for generating possible
+ * parameters
+ * @throws std::domain_error if non_random_generator_instance cannot satisfy
+ * check constraint
  */
-template <typename T1>
-void generateValues(T1& v, Constraint c, int N) {
-  v.resize(N);
-  
-  if (c == Constraint::None) {
-    for (int i = 0; i < N; i++)
-      v[i] = i - 1;
-  } else if (c == Constraint::Positive) {
-    for (int i = 0; i < N; i++)
-      v[i] = i + 1;
-  } else if (c == Constraint::NonNegative) {
-    for (int i = 0; i < N; i++)
-      v[i] = i;
+template<typename T1>
+struct generate_values {
+  template <typename T_check, typename T_generator>
+  T1 operator()(int N, T_check check,
+                T_generator &non_random_generator_instance) {
+    T1 vec(N);
+
+    for (int i = 0; i < vec.size(); i++) {
+      double v;
+
+      // There are only .size() possible non_random_numbers to try
+      for (int j = 0; j < non_random_generator_instance.size(); j++) {
+        v = non_random_generator_instance();
+        if (check(v))
+          break;
+      }
+
+      // If last number tried doesn't pass check, throw exception (constraint
+      // cannot be satisfied)
+      if (!check(v))
+        stan::math::domain_error("generate_values", "check", "",
+                                 "constraint cannot be satisfied by generator");
+
+      vec[i] = v;
+    }
+
+    return vec;
   }
-}
+};
 
 /*
- * Set v to a value that satisfies the constraint (and as much as possible
- * does not satisfy other constraints -- but this isn't totally possible)
+ * Generate a value from non_random_generator_instance that satisfies the
+ * constraint check.
+ *
+ * This is a functor instead of a function because it takes advantage of partial
+ * template specialization.
  *
  * @tparam T1 Type of v
- * @param v Variable to assign value to
- * @param c Constraint that value must fulfill
+ * @tparam T_check Callable type that takes in one double and returns a bool
+ * @tparam T_generator Callable type that takes no arguments and returns a
+ * double
  * @param N Unused
+ * @param check Callable for validating p
+ * @param non_random_generator_instance Callable for generating possible
+ * parameters
+ * @throws std::domain_error if non_random_generator_instance cannot satisfy
+ * check constraint
  */
-template <>
-void generateValues(double& v, Constraint c, int N) {
-  if (c == Constraint::None) {
-    v = -1;
-  } else if (c == Constraint::Positive) {
-    v = 1;
-  } else if (c == Constraint::NonNegative) {
-    v = 0;
+template<>
+struct generate_values<double> {
+  template <typename T_check, typename T_generator>
+  double operator()(int N, T_check check,
+                    T_generator &non_random_generator_instance) {
+    double v;
+
+    // There are only .size() possible non_random_numbers to try
+    for (int j = 0; j < non_random_generator_instance.size(); j++) {
+      v = non_random_generator_instance();
+      if (check(v))
```

Comment:
What about a paradigm where the callable just returns a number that satisfies the constraint and an orthogonal function that fills some (container | scalar) with value(s) from the generator?

Reply:
I thought about this, but got stumped at bad values, but I'm not sure why now.

I guess the simplest thing is pass in a list of good values and bad values for each parameter. That sounds simple now that I say it. I'm not sure why I didn't do it that way. Mistakes were made, sigh. I'll switch it to be like this.

This will simultaneously get rid of the non_random_generator, generate_values, and VectorizeCheck (as a side note, was looking at Google source code today and to my amusement found snake case for member functions: https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/framework/op_kernel.h#L686 -- looks like they can't quite stick to TheirStyleRecommendations)

Reply:
unrelated to this PR, are you familiar with Tensorflow? It would be awesome to figure out how to do some benchmarks against the math library for autodiff.

Reply:
Also, did you see my other comment? Github is hiding it by default, calling it outdated...? Pasting it below, from vector_rng_test_helper.hpp:172

Oh man, I think I see what you're saying - you're using the implicit knowledge of how your test parameters were generated here. Can we just make an explicit check? I think that'd be more clear and with less invisible coupling.

To explain a little more - I think the way you're passing lengths to generate_values despite the fact that only some calls will use the lengths, then later checking at runtime the lengths is a potential code smell. I might recommend playing with the API a bit more - maybe encapsulate things like the length inside of generate_values and then check what's generated on the outside. As is, the functions here rely on intimate internal knowledge of each other. And it seems a bit strange the way set_value comes in after to reset the values you just generated. I wouldn't be surprised if it had to be this complicated as you're dealing with something abstracting over a lot of different types (which is good!), but I do wonder if it could be a little simpler?

This framework seems like it's halfway to something like quickcheck - but we might be okay with hardcoding inputs of each type and constraint somewhere and trying all combinations. Go prescribes something more like the latter called table-driven testing that's simpler and maybe more appropriate here. I would find it less confusing than hiding the values you test with inside of a non_random_generator that gets passed into generate_values etc.